core/dispatch_json.rs

```diff
@@ -0,0 +1,230 @@
+    }
+    Self::Kind(kind)
+  }
+}
+
+/// Object that can be encoded to json and sent back to represent rust errors to TS/JS.
+pub trait JsonError: Sync + Send {
+  /// Returns kind represented as a JsonErrorKind. The numeric values assoiacted
+  /// to each kind are arbitrary(see explaination below).
+  ///
+  /// Their isn't a simple way to driectly share kind types between TS and rust.
+  /// It is up to the implementation to control how error kinds are mapped to
+  /// u32 values on the rust side and then converted back on the TS/JS side.
+  /// On the rust side this is pretty much free form, but I recommend using
+  /// a enum with discriminant values(see `InternalErrorKinds`). For the ts side
+  /// this is controlled by the "errorFactory":
+  /// (kind as u32, message) -> errorFactory -> Error
+  fn kind(&self) -> JsonErrorKind;
+  /// Error message as string value.
+  fn msg(&self) -> String;
+}
+
+/// Internal error kinds for json dispatch system
+/// These values are serialized as negative(x * -1)!
+// Warning! The values in this enum are duplicated in dispatch_json.ts
+// Update carefully
+#[derive(Clone, Copy, PartialEq, Debug)]
+enum InternalErrorKinds {
+  JsonIoError = 1,
+  JsonSyntaxError = 2,
+  JsonDataError = 3,
+  JsonEofError = 4,
+}
+
+/// Internal error expression and conversion implementation.
+trait InternalError {
+  /// Kind is specific to implementation.
+  fn kind(&self) -> JsonErrorKind;
+  /// Error message as string value.
+  fn msg(&self) -> String;
+  /// Make InternalJsonError from this object
+  fn to_final_json_error(&self) -> FinalJsonError {
+    let kind = match self.kind() {
+      JsonErrorKind::Kind(k) => {
+        assert!(k != 0, "kind = 0 is reserved for UnKind");
+        k as i64 * -1
+      }
+      JsonErrorKind::UnKind => 0,
+    };
+    FinalJsonError {
+      kind,
+      message: self.msg(),
+    }
+  }
+}
+
+impl InternalError for serde_json::error::Error {
+  fn kind(&self) -> JsonErrorKind {
+    use serde_json::error::Category::*;
+    JsonErrorKind::Kind(match self.classify() {
+      Io => InternalErrorKinds::JsonIoError,
+      Syntax => InternalErrorKinds::JsonSyntaxError,
+      Data => InternalErrorKinds::JsonDataError,
+      Eof => InternalErrorKinds::JsonEofError,
+    } as u32)
+  }
+
+  fn msg(&self) -> String {
+    self.to_string()
+  }
+}
+
+/// Internal representation of an error from inside or outside of json dispatch.
+/// All errors get converted to this format before being serialized.
+#[derive(Debug, Clone, Serialize)]
+struct FinalJsonError {
+  pub kind: i64,
+  pub message: String,
+}
+
+impl<E: JsonError + Send + Sync> From<E> for FinalJsonError {
+  fn from(e: E) -> FinalJsonError {
+    let kind = match e.kind() {
+      JsonErrorKind::Kind(k) => {
+        assert!(k != 0, "kind = 0 is reserved for UnKind");
+        k as i64
+      }
+      JsonErrorKind::UnKind => 0,
+    };
+    FinalJsonError {
+      kind,
+      message: e.msg(),
+    }
+  }
+}
```

Comment:
This is really complex, PTAL at my comment below, I believe we could get rid of it completely

Reply:
This is internal only. Not relevant to implementations of `dispatch_json`. The existing implementation in cli only requires a single trait implementation for `OpError`:
``` rust
impl JsonError for OpError {
  fn kind(&self) -> JsonErrorKind {
    (self.kind as u32).into()
  }

  fn msg(&self) -> String {
    self.msg.clone()
  }
}
```